File: java-client/src/main/java/com/openbayes/client/RunWorkspaceTaskExample.java
package com.openbayes.client;

import com.openbayes.client.model.JobResult;
import com.openbayes.client.model.OpenBayesException;
import com.openbayes.client.model.WorkspaceTaskSpec;

/**
 * Clone mode: run a task that reuses a workspace you prepared earlier — no code upload.
 *
 * <p>The customer prepares a workspace (debugs code, puts it under {@code /output}), then every
 * task just binds that workspace's output and runs a command. No source code, no upload.
 */
public final class RunWorkspaceTaskExample {

    public static void main(String[] args) {
        Config cfg = Config.fromEnv();

        GraphQLClient gql = new GraphQLClient(cfg.graphqlEndpoint, cfg.token);
        OpenBayesClient client = new OpenBayesClient(gql);

        if (isBlank(cfg.token)) {
            System.out.println("使用账号密码登录获取 token...");
            client.login(cfg.username, cfg.password);
        }

        System.out.println("创建 task（绑定已准备好的工作空间 " + cfg.workspaceId + " 的 /output，零上传）...");
        WorkspaceTaskSpec spec = new WorkspaceTaskSpec(
                cfg.projectId, cfg.workspaceId, cfg.runtime, cfg.resource, cfg.command);
        JobResult job = client.createTaskFromWorkspace(cfg.party, spec);

        System.out.println("✅ 任务创建成功 jobId = " + job.id());
        String frontend = job.link("frontend");
        if (frontend != null) {
            System.out.println("   " + frontend);
        }
    }

    private static boolean isBlank(String s) {
        return s == null || s.isBlank();
    }

    /** Reads configuration from environment variables and validates required ones. */
    static final class Config {
        String graphqlEndpoint;
        String token;
        String username;
        String password;
        String party;
        String projectId;
        String workspaceId;
        String resource;
        String runtime;
        String command;

        static Config fromEnv() {
            Config c = new Config();
            c.graphqlEndpoint = env("OPENBAYES_GRAPHQL", "https://openbayes.com/gateway");
            c.token = System.getenv("OPENBAYES_TOKEN");
            c.username = System.getenv("OPENBAYES_USERNAME");
            c.password = System.getenv("OPENBAYES_PASSWORD");
            c.party = require("OPENBAYES_PARTY");
            c.projectId = require("OPENBAYES_PROJECT_ID");        // 工作空间所在的项目
            c.workspaceId = require("OPENBAYES_WORKSPACE_ID");    // 客户预先准备好的工作空间 id
            c.resource = env("OPENBAYES_RESOURCE", "cpu");
            c.runtime = require("OPENBAYES_RUNTIME");             // 建议与工作空间一致
            c.command = require("OPENBAYES_COMMAND");             // 如 python /output/main.py

            boolean hasToken = c.token != null && !c.token.isBlank();
            boolean hasLogin = c.username != null && !c.username.isBlank()
                    && c.password != null && !c.password.isBlank();
            if (!hasToken && !hasLogin) {
                throw new OpenBayesException(
                        "缺少鉴权: 请设置 OPENBAYES_TOKEN，或同时设置 OPENBAYES_USERNAME 和 OPENBAYES_PASSWORD");
            }
            return c;
        }

        private static String env(String name, String fallback) {
            String v = System.getenv(name);
            return (v == null || v.isBlank()) ? fallback : v;
        }

        private static String require(String name) {
            String v = System.getenv(name);
            if (v == null || v.isBlank()) {
                throw new OpenBayesException("缺少必需的环境变量: " + name);
            }
            return v;
        }
    }
}

File: python-client/run_workspace_task_example.py
"""Clone mode: run a task that reuses a workspace you prepared earlier — no code upload.

The customer prepares a workspace (debugs code, puts it under /output), then every task just
binds that workspace's output and runs a command. No source code, no upload.

    [token or login] -> create_task_from_workspace(bind workspace /output -> /output, run command)

Run it from the python-client/ directory:  python run_workspace_task_example.py
"""
import os
import sys

from openbayes_client import (
    GraphQLClient,
    OpenBayesClient,
    OpenBayesError,
    WorkspaceTaskSpec,
)


def _env(name, fallback=None):
    v = os.environ.get(name)
    return v if v else fallback


def _require(name):
    v = os.environ.get(name)
    if not v:
        raise OpenBayesError("缺少必需的环境变量: " + name)
    return v


def main():
    endpoint = _env("OPENBAYES_GRAPHQL", "https://openbayes.com/gateway")
    token = _env("OPENBAYES_TOKEN")
    username = _env("OPENBAYES_USERNAME")
    password = _env("OPENBAYES_PASSWORD")
    party = _require("OPENBAYES_PARTY")
    project_id = _require("OPENBAYES_PROJECT_ID")      # 工作空间所在的项目
    workspace_id = _require("OPENBAYES_WORKSPACE_ID")  # 客户预先准备好的工作空间 id
    resource = _env("OPENBAYES_RESOURCE", "cpu")
    runtime = _require("OPENBAYES_RUNTIME")            # 建议与工作空间一致
    command = _require("OPENBAYES_COMMAND")            # 如 python /output/main.py

    if not token and not (username and password):
        raise OpenBayesError(
            "缺少鉴权: 请设置 OPENBAYES_TOKEN，或同时设置 OPENBAYES_USERNAME 和 OPENBAYES_PASSWORD"
        )

    gql = GraphQLClient(endpoint, token)
    client = OpenBayesClient(gql)

    if not token:
        print("使用账号密码登录获取 token...")
        client.login(username, password)

    print("创建 task（绑定已准备好的工作空间 {} 的 /output，零上传）...".format(workspace_id))
    spec = WorkspaceTaskSpec(
        project_id=project_id,
        workspace_id=workspace_id,
        runtime=runtime,
        resource=resource,
        command=command,
    )
    job = client.create_task_from_workspace(party, spec)

    print("✅ 任务创建成功 jobId = " + job.id)
    frontend = job.link("frontend")
    if frontend:
        print("   " + frontend)


if __name__ == "__main__":
    try:
        main()
    except OpenBayesError as e:
        print("错误: {}".format(e), file=sys.stderr)
        sys.exit(1)
